File: C++/Aula 04/ex1.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Livro{
    protected:
        int chave;
        string titulo;
        string autor;
        int ano;

    private:
        static int proximaChave;

    public:
        Livro(const string& titulo, const string& autor, int ano)
            : titulo(titulo), autor(autor), ano(ano){
            this->chave = proximaChave++;
            cout << "[+] Livro criado (base): \"" << this->titulo << "\" (chave " << this->chave << ")\n";
        }

        virtual ~Livro(){
            cout << "[-] Livro destruído: \"" << titulo << "\" (chave " << chave << ")\n";
        }

        virtual void exibir() const{
            cout << "  Chave : " << chave  << "\n";
            cout << "  Título: " << titulo << "\n";
            cout << "  Autor : " << autor  << "\n";
            cout << "  Ano   : " << (ano > 0 ? to_string(ano) : "N/A") << "\n";
        }
};

int Livro::proximaChave = 1;

class LivroFisico : public Livro{
    private:
        int numeroPaginas;

    public:
        LivroFisico(const string& titulo, const string& autor, int ano, int numeroPaginas)
            : Livro(titulo, autor, ano), numeroPaginas(numeroPaginas){
            cout << "[+] LivroFisico criado: \"" << this->titulo << "\" (" << numeroPaginas << " págs.)\n";
        }

        ~LivroFisico(){
            cout << "[-] LivroFisico destruído: \"" << titulo << "\"\n";
        }

        void exibir() const override {
            cout << "  ---- Livro Físico ----\n";
            Livro::exibir();
            cout << "  Págs. : " << numeroPaginas << "\n";
        }
};

class LivroDigital : public Livro{
    private:
        float tamanhoArquivo;
        string formato;

    public:
        LivroDigital(const string& titulo, const string& autor, int ano,
                    float tamanhoArquivo, const string& formato)
            : Livro(titulo, autor, ano), tamanhoArquivo(tamanhoArquivo), formato(formato){
            cout << "[+] LivroDigital criado: \"" << this->titulo << "\" (" << formato << ", " << tamanhoArquivo << " MB)\n";
        }

        ~LivroDigital(){
            cout << "[-] LivroDigital destruído: \"" << titulo << "\"\n";
        }

        void exibir() const override{
            cout << "  ---- Livro Digital ----\n";
            Livro::exibir();
            cout << "  Formato: " << formato << "\n";
            cout << "  Tamanho: " << tamanhoArquivo << " MB\n";
        }
};

class LivroAcademicoDigital : public LivroDigital{
    private:
        string area;
        string instituicao;

    public:
        LivroAcademicoDigital(const string& titulo, const string& autor, int ano,
                            float tamanhoArquivo, const string& formato,
                            const string& area, const string& instituicao)
            : LivroDigital(titulo, autor, ano, tamanhoArquivo, formato),
            area(area), instituicao(instituicao){
            cout << "[+] LivroAcademicoDigital criado: \"" << this->titulo << "\" | " << area << " - " << instituicao << "\n";
        }

        ~LivroAcademicoDigital(){
            cout << "[-] LivroAcademicoDigital destruído: \"" << titulo << "\"\n";
        }

        void exibir() const override{
            cout << "  ---- Livro Acadêmico Digital ----\n";
            LivroDigital::exibir();
            cout << "  Área  : " << area        << "\n";
            cout << "  Inst. : " << instituicao << "\n";
        }
};

int main(){
    cout << "=== Criando objetos ===\n\n";

    vector<Livro*> colecao;
    colecao.push_back(new Livro("Dom Casmurro", "Machado de Assis", 1899));
    colecao.push_back(new LivroFisico("O Senhor dos Anéis", "J.R.R. Tolkien", 1954, 1178));
    colecao.push_back(new LivroDigital("Clean Code", "Robert C. Martin", 2008, 2.5f, "PDF"));
    colecao.push_back(new LivroAcademicoDigital(
        "Introdução à Computação", "Ana Silva", 2020,
        1.8f, "EPUB", "Ciência da Computação", "UTFPR"
    ));

    cout << "\n=== Exibindo coleção ===\n";
    for (const auto& livro : colecao){
        cout << "\n";
        livro->exibir();
    }

    cout << "\n=== Destruindo objetos (ordem reversa) ===\n\n";
    for (int i = colecao.size() - 1; i >= 0; i--){
        delete colecao[i];
        cout << "\n";
    }
    colecao.clear();

    return 0;
}

File: Python/Aula 04/ex1.py
class Livro:
    # Atributo classe para controlar a chave incremental
    _proxima_chave = 1 

    def __init__(self, titulo: str, autor: str, ano: int):
        # A chave recebe o valor atual e incrementa para o próximo cadastrado
        self.chave = Livro._proxima_chave
        Livro._proxima_chave += 1
        
        self.titulo = titulo
        self.autor = autor
        self.ano = ano

    def __del__(self):
        # Destrutor, indica que o objeto foi desalocado
        print(f"[Destrutor] O objeto Livro '{self.titulo}' (Chave: {self.chave}) foi desalocado da memória.")

    def exibir_informacoes(self):
        print(f"[Livro Base] Chave: {self.chave} | Título: {self.titulo} | Autor: {self.autor} | Ano: {self.ano}")


class LivroFisico(Livro):
    def __init__(self, titulo: str, autor: str, ano: int, numero_paginas: int):
        # Utiliza o construtor da classe base (super)
        super().__init__(titulo, autor, ano)
        self.numero_paginas = numero_paginas

    def exibir_informacoes(self):
        # Sobrescreve método (Polimorfismo) para exibir informações completas
        print(f"[Livro Físico] Chave: {self.chave} | Título: {self.titulo} | Autor: {self.autor} | Ano: {self.ano} | Páginas: {self.numero_paginas}")


class LivroDigital(Livro):
    def __init__(self, titulo: str, autor: str, ano: int, tamanho_arquivo: float, formato: str):
        # Utiliza o construtor da classe base (super)
        super().__init__(titulo, autor, ano)
        self.tamanho_arquivo = tamanho_arquivo
        self.formato = formato

    def exibir_informacoes(self):
        # Sobrescreve método (Polimorfismo) para exibir informações completas
        print(f"[Livro Digital] Chave: {self.chave} | Título: {self.titulo} | Autor: {self.autor} | Ano: {self.ano} | Tamanho: {self.tamanho_arquivo} MB | Formato: {self.formato}")


class LivroAcademicoDigital(LivroDigital):
    def __init__(self, titulo: str, autor: str, ano: int, tamanho_arquivo: float, formato: str, area: str, instituicao: str):
        # Utiliza o construtor da classe LivroDigital que repassa classe Livro
        super().__init__(titulo, autor, ano, tamanho_arquivo, formato)
        self.area = area
        self.instituicao = instituicao

    def exibir_informacoes(self):
        # Sobrescreve método (Polimorfismo) para exibir informações completas
        print(f"[Livro Acadêmico Digital] Chave: {self.chave} | Título: {self.titulo} | Autor: {self.autor} | Ano: {self.ano} | Tamanho: {self.tamanho_arquivo} MB | Formato: {self.formato} | Área: {self.area} | Instituição: {self.instituicao}")


if __name__ == '__main__':
    print("INICIANDO SISTEMA DE LIVROS\n")

    # Cria uma lista contendo livros de diferentes tipos para demonstrar o Polimorfismo
    colecao_livros = [
        Livro("Dom Casmurro", "Machado de Assis", 1899),
        LivroFisico("O Senhor dos Anéis", "J.R.R. Tolkien", 1954, 1200),
        LivroDigital("Python Fluente", "Luciano Ramalho", 2015, 15.5, "PDF"),
        LivroAcademicoDigital("Estrutura de Dados e Algoritmos", "Thomas H. Cormen", 2009, 25.0, "EPUB", "Ciência da Computação", "MIT")
    ]

    # Percorre a coleção e exibe informações
    print("EXIBINDO INFORMAÇÕES DOS LIVROS")
    for livro in colecao_livros:
        livro.exibir_informacoes()
        
    print("\nFIM DO PROGRAMA (Iniciando desalocação)")